mozilla/telemetry-airflow
# feat!: Use Dockerfile based on upstream [DSRE-1116]

### ⚠ ~Do not merge to main, this is not compatible with existing deployment config, as it will require changes to the Helm chart~ ⚠ See related PR for Helm chart changes

# Description
Extend Airflow official docker image instead of baking our own image. 


## Highlights
* CI runtime decreased by 70-80%
* docker-compose setup time decreased significantly. **Local use: takes a few seconds as opposed to a few minutes before**. Airflow Variables and Connections are loaded via `import` CLI command; replaces `bin/run` script using Airflow CLI.
* Improved dev secrets security by dynamically generating a Fernet key in makefile
  * `.env` file is generated from `make up`, contains UID and Fernet key. `.env` is automatically loaded into environment variables by `docker-compose`
* Retire shell scripts to use builtin features in our stack
  * `bin/run` replaced by docker-compose and CI
  * `bin/test-dag-tags` replaced by pytest unit tests
  * `bin/test-parse` replaced by pytest unit tests
  * This will likely require changes to bqetl for DAG validation

## Dockerfile Changes
Dockerfile now based on [upstream Airflow image](https://hub.docker.com/r/apache/airflow)

- Local port 8000 ➡️ 8080
- Default user ID 10000 ➡️ 50000 to match Docker image
  - Official image implements [arbitrary docker user](https://airflow.apache.org/docs/docker-stack/entrypoint.html#allowing-arbitrary-user-to-run-the-container) with root `GID` . Actual Docker UID is set dynamically at runtime and entrypoint script ensures it is provisioned appropriately

### Open questions
- [x] Review dependencies installed in the original Dockerfile, are they needed? 👇
```Dockerfile
RUN apt-get update && \
    apt-get install -y --no-install-recommends \
        apt-transport-https build-essential curl git libpq-dev python-dev \
        default-libmysqlclient-dev gettext sqlite3 libffi-dev \
        lsb-release gnupg vim screen procps default-mysql-client && \
    CLOUD_SDK_REPO="cloud-sdk-$(lsb_release -c -s)" && \
    echo "deb http://packages.cloud.google.com/apt $CLOUD_SDK_REPO main" | tee -a /etc/apt/sources.list.d/google-cloud-sdk.list && \
    curl https://packages.cloud.google.com/apt/doc/apt-key.gpg | apt-key add - && \
    apt-get update -y && apt-get install google-cloud-sdk -y && apt-get install google-cloud-sdk-gke-gcloud-auth-plugin && \
    apt-get remove -y lsb-release gnupg && \
```

## Docker-compose.yaml changes
docker-compose.yaml based on upstream [docker-compose configuration](https://airflow.apache.org/docs/apache-airflow/stable/howto/docker-compose/index.html#fetching-docker-compose-yaml). This includes the `airflow-init` which handles database migrations and setting up other services in the right order.

-  bin/run removed
  - Service initialization is handled by `docker-compose` service `airflow-init`
  - Connections and variables are batch imported by CLI to speed up Airflow initialization
- `AIRFLOW_UID` inserted into environment variables in `docker-compose` instead of needing a specific build command for linux users
- add health checks for services

## CI changes
- Directly run tests instead of using shell scripts
- Replace integration tests by unit tests

## Others
#### Dependencies
- removed crypto Airflow extra ([deprecated](https://airflow.apache.org/docs/apache-airflow/stable/extra-packages-ref.html#deprecated-1-10-extras))
#### Makefile
- removed `build-linux` since `AIRFLOW_UID` is handled by docker-compose
- Dynamically generate fernet key in `make up`
#### Unit tests
- test dag validity
- check required  tags for each dag
- check valid tags for each dag

# TODO

- [x] Resolve all open questions

# Related PR
https://github.com/mozilla-services/cloudops-infra/pull/4705

# Related Jira Ticket
* [DSRE-1116](https://mozilla-hub.atlassian.net/browse/DSRE-1116)

[DSRE-1116]: https://mozilla-hub.atlassian.net/browse/DSRE-1116?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ
